{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3815742b-f83f-4b72-bff0-1393424ef961",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "from tqdm import tqdm\n",
    "import requests\n",
    "import time\n",
    "import os\n",
    "import shutil\n",
    "import matplotlib.pyplot as plt\n",
    "import sys\n",
    "import platform\n",
    "\n",
    "def load_manifest_file(location):\n",
    "    \"\"\"\n",
    "    Function use:\n",
    "    Load the GDC manifest file from a specified directory.\n",
    "\n",
    "    Inputs:\n",
    "    - location (str): Directory path containing the GDC manifest file.\n",
    "\n",
    "    Outputs:\n",
    "    - DataFrame: Contains the data from the manifest file if found.\n",
    "    - None: If no matching file is found.\n",
    "    \"\"\"\n",
    "    pattern = f\"{location}/*gdc_manifest*\"\n",
    "    files = glob.glob(pattern)\n",
    "    if files:\n",
    "        return pd.read_csv(files[0], sep='\\t')\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "def fetch_sample_type_from_file_uuid(file_uuid):\n",
    "    \"\"\"\n",
    "    Function use:\n",
    "    Fetch the sample type for a given file UUID using the GDC API.\n",
    "\n",
    "    Inputs:\n",
    "    - file_uuid (str): UUID of the file.\n",
    "\n",
    "    Outputs:\n",
    "    - sample_type (str): Sample type (e.g., 'Primary Tumor').\n",
    "    - None: If the API request fails.\n",
    "    \"\"\"\n",
    "    base_url = 'https://api.gdc.cancer.gov'\n",
    "    response = requests.get(f'{base_url}/files/{file_uuid}?expand=cases.samples')\n",
    "    if response.status_code == 200:\n",
    "        data = response.json()\n",
    "        return data.get('data', {}).get('cases', [{}])[0].get('samples', [{}])[0].get('sample_type', 'Not Found')\n",
    "    else:\n",
    "        print(f\"Failed to retrieve sample type for UUID {file_uuid}.\")\n",
    "        return None\n",
    "\n",
    "def fetch_diagnosis_details_from_file_uuid(file_uuid):\n",
    "    \"\"\"\n",
    "    Function use:\n",
    "    Fetch diagnosis details for a given file UUID using the GDC API.\n",
    "\n",
    "    Inputs:\n",
    "    - file_uuid (str): UUID of the file.\n",
    "\n",
    "    Outputs:\n",
    "    - diagnosis_details (list): List of dictionaries containing diagnosis information.\n",
    "    - Empty list: If the API request fails.\n",
    "    \"\"\"\n",
    "    base_url = 'https://api.gdc.cancer.gov'\n",
    "    response = requests.get(f'{base_url}/files/{file_uuid}?expand=cases.diagnoses')\n",
    "    if response.status_code == 200:\n",
    "        data = response.json()\n",
    "        return data.get('data', {}).get('cases', [{}])[0].get('diagnoses', [])\n",
    "    else:\n",
    "        print(f\"Failed to retrieve diagnosis details for UUID {file_uuid}.\")\n",
    "        return []\n",
    "\n",
    "def fetch_clinical_data_for_manifest(manifest, n=None, t=1):\n",
    "    \"\"\"\n",
    "    Function use:\n",
    "    Enrich a manifest DataFrame with sample types and diagnosis details fetched from the GDC API.\n",
    "\n",
    "    Inputs:\n",
    "    - manifest (DataFrame): Manifest with file UUIDs in an 'id' column.\n",
    "    - n (int): Maximum number of files to process. If None, all rows are processed.\n",
    "    - t (float): Time (in seconds) to wait between API calls.\n",
    "\n",
    "    Outputs:\n",
    "    - Updated DataFrame: Includes additional sample type and diagnosis details.\n",
    "    \"\"\"\n",
    "    if 'sample_type' not in manifest.columns:\n",
    "        manifest['sample_type'] = None\n",
    "\n",
    "    if n is None:\n",
    "        n = len(manifest)\n",
    "\n",
    "    for i in tqdm(range(min(n, len(manifest)))):\n",
    "        if i > 0:\n",
    "            time.sleep(t)\n",
    "        file_id = manifest.loc[i, 'id']\n",
    "        sample_type = fetch_sample_type_from_file_uuid(file_id)\n",
    "        if sample_type:\n",
    "            manifest.at[i, 'sample_type'] = sample_type\n",
    "        else:\n",
    "            print(f\"No sample type found for file UUID {file_id}.\")\n",
    "        diagnosis_details = fetch_diagnosis_details_from_file_uuid(file_id)\n",
    "        if diagnosis_details and isinstance(diagnosis_details, list):\n",
    "            for detail in diagnosis_details:\n",
    "                if isinstance(detail, dict):\n",
    "                    for key, value in detail.items():\n",
    "                        column_name = f\"diagnosis_{key}\"\n",
    "                        if column_name not in manifest.columns:\n",
    "                            manifest[column_name] = None\n",
    "                        manifest.at[i, column_name] = value\n",
    "                else:\n",
    "                    print(f\"Detail is not a dictionary: {detail}\")\n",
    "        else:\n",
    "            print(f\"Diagnosis details structure unexpected or missing for file ID {file_id}: {diagnosis_details}\")\n",
    "\n",
    "    return manifest\n",
    "\n",
    "def save_clinical_manifest(manifest, directory):\n",
    "    \"\"\"\n",
    "    Function use:\n",
    "    Save the clinical manifest DataFrame to a file.\n",
    "\n",
    "    Inputs:\n",
    "    - manifest (DataFrame): Data to save.\n",
    "    - directory (str): Directory path where the file will be saved.\n",
    "\n",
    "    Outputs:\n",
    "    - None: File is saved directly to the specified directory.\n",
    "    \"\"\"\n",
    "    directory = os.path.join(directory, '')\n",
    "    filepath = f\"{directory}clinical_gdc_manifest.txt\"\n",
    "    manifest.to_csv(filepath, sep='\\t', index=False)\n",
    "    print(f\"File saved successfully at {filepath}\")\n",
    "\n",
    "def load_clinical_manifest(directory):\n",
    "    \"\"\"\n",
    "    Function use:\n",
    "    Load a clinical manifest file into a DataFrame.\n",
    "\n",
    "    Inputs:\n",
    "    - directory (str): Directory path where the file is located.\n",
    "\n",
    "    Outputs:\n",
    "    - DataFrame: Contains the clinical manifest data.\n",
    "    \"\"\"\n",
    "    filepath = os.path.join(directory, 'clinical_gdc_manifest.txt')\n",
    "    clinical_manifest = pd.read_csv(filepath, sep='\\t')\n",
    "    print('Clinical manifest shape:', clinical_manifest.shape)\n",
    "    return clinical_manifest\n",
    "\n",
    "def clean_dataframe(df):\n",
    "    \"\"\"\n",
    "    Function use:\n",
    "    Clean a DataFrame by removing columns with irrelevant or missing data.\n",
    "\n",
    "    Inputs:\n",
    "    - df (DataFrame): DataFrame to clean.\n",
    "\n",
    "    Outputs:\n",
    "    - DataFrame: Cleaned version of the input DataFrame.\n",
    "    \"\"\"\n",
    "    df_copy = df.copy()\n",
    "    for column in df_copy.columns:\n",
    "        unique_values = df_copy[column].dropna().unique()\n",
    "        if len(unique_values) == 0 or (len(unique_values) == 1 and unique_values[0] in ['Not Reported', 'not reported', 'None']):\n",
    "            df_copy.drop(column, axis=1, inplace=True)\n",
    "        else:\n",
    "            if column in [\"ajcc_pathologic_stage\", \"tumor_grade\", \"ajcc_pathologic_t\", \"ajcc_pathologic_m\",\n",
    "                          \"ajcc_pathologic_n\", \"figo_stage\", \"wilms_tumor_histologic_subtype\"]:\n",
    "                df_copy.loc[df_copy['sample_type'] == 'Solid Tissue Normal', column] = 'healthy tissue'\n",
    "    print('Cleaned clinical manifest shape:', df_copy.shape)\n",
    "    return df_copy\n",
    "\n",
    "def organize_tsv_files(base_dir):\n",
    "    \"\"\"\n",
    "    Function use:\n",
    "    Organize .tsv files by moving them from subdirectories to the base directory.\n",
    "\n",
    "    Inputs:\n",
    "    - base_dir (str): Path to the base directory.\n",
    "\n",
    "    Outputs:\n",
    "    - None: Files are moved, and subdirectories are cleaned.\n",
    "    \"\"\"\n",
    "    has_directories = False\n",
    "    for root, dirs, files in os.walk(base_dir, topdown=False):\n",
    "        if dirs:\n",
    "            has_directories = True\n",
    "        for file in files:\n",
    "            if file.endswith(\".tsv\"):\n",
    "                shutil.move(os.path.join(root, file), os.path.join(base_dir, file))\n",
    "    if has_directories:\n",
    "        for root, dirs, files in os.walk(base_dir, topdown=False):\n",
    "            for dir in dirs:\n",
    "                shutil.rmtree(os.path.join(root, dir))\n",
    "        print(\"Cleaning done\")\n",
    "    else:\n",
    "        print(\"No directories to clean. Cleaning done.\")\n",
    "\n",
    "def load_gene_expression(manifest_df, exp_dir, feature):\n",
    "    \"\"\"\n",
    "    Function use:\n",
    "    Load gene expression data from .tsv files, keeping only relevant rows and columns.\n",
    "\n",
    "    Inputs:\n",
    "    - manifest_df (DataFrame): Manifest with filenames.\n",
    "    - exp_dir (str): Directory containing .tsv files.\n",
    "    - feature (str): Gene expression feature to extract.\n",
    "\n",
    "    Outputs:\n",
    "    - gene_expression_df (DataFrame): Contains gene expression data.\n",
    "    - gene_key_df (DataFrame): Maps gene IDs to gene names.\n",
    "    \"\"\"\n",
    "    gene_expression_data = {}\n",
    "    gene_key = {}\n",
    "\n",
    "    for filename in manifest_df['filename']:\n",
    "        file_path = os.path.join(exp_dir, filename)\n",
    "        try:\n",
    "            df = pd.read_csv(file_path, sep='\\t', comment='#')\n",
    "            df = df[df['gene_id'].str.contains('ENSG')]\n",
    "            gene_expression_data[filename] = df[feature].values\n",
    "            gene_key.update(dict(zip(df['gene_id'], df['gene_name'])))\n",
    "        except Exception as e:\n",
    "            print(f\"Failed to load {filename}: {e}\")\n",
    "\n",
    "    gene_expression_df = pd.DataFrame(gene_expression_data, index=list(gene_key.keys()))\n",
    "    gene_expression_df.index.name = 'gene_id'\n",
    "    gene_expression_df.columns.name = 'filename'\n",
    "\n",
    "    gene_key_df = pd.DataFrame(list(gene_key.items()), columns=['gene_id', 'gene_name'])\n",
    "\n",
    "    print('Gene expression shape:', gene_expression_df.T.shape)\n",
    "    print('Gene key shape:', gene_key_df.shape)\n",
    "\n",
    "    return gene_expression_df.T, gene_key_df\n",
    "\n",
    "def save_dataframes(gene_exps, gene_key, cancer):\n",
    "    \"\"\"\n",
    "    Function use:\n",
    "    Save gene expression and key DataFrames to files.\n",
    "\n",
    "    Inputs:\n",
    "    - gene_exps (DataFrame): Gene expression data.\n",
    "    - gene_key (DataFrame): Gene key data.\n",
    "    - cancer (str): Directory for saving the data.\n",
    "\n",
    "    Outputs:\n",
    "    - None: DataFrames are saved to files.\n",
    "    \"\"\"\n",
    "    gene_exps.to_csv(f'./data/TCGA_GeneExpression/{cancer}/gene_expression.csv', index=True)\n",
    "    gene_key.to_csv(f'./data/TCGA_GeneExpression/{cancer}/gene_key.csv', index=True)\n",
    "\n",
    "def load_dataframes(exps_filename, key_filename):\n",
    "    \"\"\"\n",
    "    Function use:\n",
    "    Load gene expression and key DataFrames from files.\n",
    "\n",
    "    Inputs:\n",
    "    - exps_filename (str): Path to gene expression file.\n",
    "    - key_filename (str): Path to gene key file.\n",
    "\n",
    "    Outputs:\n",
    "    - gene_exps (DataFrame): Loaded gene expression data.\n",
    "    - gene_key (DataFrame): Loaded gene key data.\n",
    "    \"\"\"\n",
    "    gene_exps = pd.read_csv(exps_filename, index_col=0)\n",
    "    gene_key = pd.read_csv(key_filename, index_col=0)\n",
    "    return gene_exps, gene_key\n",
    "\n",
    "def analyze_filenames(manifest_df, directory_path):\n",
    "    \"\"\"\n",
    "    Analyze the uniqueness and overlap of filenames between a manifest and a directory.\n",
    "\n",
    "    Inputs:\n",
    "    - manifest_df (DataFrame): Manifest containing filenames.\n",
    "    - directory_path (str): Directory path containing files to compare.\n",
    "\n",
    "    Outputs:\n",
    "    - dict: Summary of percentages including unique filenames, overlap, and non-overlap.\n",
    "    \"\"\"\n",
    "    total_filenames = len(manifest_df['filename'])\n",
    "    unique_filenames = len(manifest_df['filename'].unique())\n",
    "    percent_unique = (unique_filenames / total_filenames) * 100\n",
    "\n",
    "    directory_files = set(os.listdir(directory_path))\n",
    "    manifest_files = set(manifest_df['filename'])\n",
    "\n",
    "    overlap_files = manifest_files.intersection(directory_files)\n",
    "    percent_overlap = (len(overlap_files) / total_filenames) * 100\n",
    "    non_overlap_files = manifest_files.difference(directory_files)\n",
    "    percent_non_overlap = (len(non_overlap_files) / total_filenames) * 100\n",
    "\n",
    "    return {\n",
    "        \"percent_unique_filenames_in_manifest\": percent_unique,\n",
    "        \"percent_overlap_with_directory\": percent_overlap,\n",
    "        \"percent_non_overlap_with_directory\": percent_non_overlap\n",
    "    }\n",
    "\n",
    "# Show Python version, platform details, and versions of all imported packages.\n",
    "def environment_info():\n",
    "    \"\"\"\n",
    "    Display Python version, platform details, and versions of all imported packages.\n",
    "    \"\"\"\n",
    "    # Python and system information\n",
    "    print(\"Python version:\", sys.version)\n",
    "    print(\"Platform:\", platform.platform())\n",
    "\n",
    "    # Imported package versions\n",
    "    print(\"\\nPackage Versions:\")\n",
    "    imported_modules = {name: module.__version__ for name, module in sys.modules.items() \n",
    "                        if hasattr(module, '__version__')}\n",
    "    for package, version in sorted(imported_modules.items()):\n",
    "        print(f\"{package}: {version}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fa7c2ad6-b6af-4b67-ae56-d60ed2fad885",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python version: 3.12.2 | packaged by conda-forge | (main, Feb 16 2024, 21:00:12) [Clang 16.0.6 ]\n",
      "Platform: macOS-14.5-x86_64-i386-64bit\n",
      "\n",
      "Package Versions:\n",
      "IPython: 8.25.0\n",
      "IPython.core.release: 8.25.0\n",
      "PIL: 10.3.0\n",
      "PIL.Image: 10.3.0\n",
      "PIL._deprecate: 10.3.0\n",
      "PIL._version: 10.3.0\n",
      "_brotli: 1.0.9\n",
      "_csv: 1.0\n",
      "_ctypes: 1.1.0\n",
      "_curses: b'2.2'\n",
      "_decimal: 1.70\n",
      "_pydev_bundle.fsnotify: 0.1.5\n",
      "_pydevd_frame_eval.vendored.bytecode: 0.13.0.dev\n",
      "appnope: 0.1.4\n",
      "argparse: 1.1\n",
      "brotli: 1.0.9\n",
      "certifi: 2024.02.02\n",
      "cffi: 1.16.0\n",
      "charset_normalizer: 2.0.4\n",
      "charset_normalizer.version: 2.0.4\n",
      "comm: 0.2.2\n",
      "csv: 1.0\n",
      "ctypes: 1.1.0\n",
      "ctypes.macholib: 1.0\n",
      "cycler: 0.12.1\n",
      "dateutil: 2.9.0\n",
      "dateutil._version: 2.9.0\n",
      "debugpy: 1.6.7\n",
      "debugpy.public_api: 1.6.7\n",
      "decimal: 1.70\n",
      "decorator: 5.1.1\n",
      "defusedxml: 0.7.1\n",
      "executing: 2.0.1\n",
      "executing.version: 2.0.1\n",
      "http.server: 0.6\n",
      "idna: 3.4\n",
      "idna.idnadata: 15.0.0\n",
      "idna.package_data: 3.4\n",
      "ipaddress: 1.0\n",
      "ipykernel: 6.29.3\n",
      "ipykernel._version: 6.29.3\n",
      "jedi: 0.19.1\n",
      "json: 2.0.9\n",
      "jupyter_client: 8.6.2\n",
      "jupyter_client._version: 8.6.2\n",
      "jupyter_core: 5.7.2\n",
      "jupyter_core.version: 5.7.2\n",
      "kiwisolver: 1.4.5\n",
      "kiwisolver._cext: 1.4.5\n",
      "logging: 0.5.1.2\n",
      "matplotlib: 3.9.0\n",
      "numpy: 1.26.4\n",
      "numpy._core._multiarray_umath: 3.1\n",
      "numpy.core: 1.26.4\n",
      "numpy.core._multiarray_umath: 3.1\n",
      "numpy.linalg._umath_linalg: 0.1.5\n",
      "numpy.version: 1.26.4\n",
      "packaging: 24.0\n",
      "pandas: 2.2.2\n",
      "pandas._version_meson: 2.2.2\n",
      "parso: 0.8.4\n",
      "pickleshare: 0.7.5\n",
      "pkg_resources._vendor.more_itertools: 9.1.0\n",
      "pkg_resources._vendor.packaging: 23.1\n",
      "pkg_resources._vendor.platformdirs: 2.6.2\n",
      "pkg_resources._vendor.platformdirs.version: 2.6.2\n",
      "pkg_resources.extern.more_itertools: 9.1.0\n",
      "pkg_resources.extern.packaging: 23.1\n",
      "pkg_resources.extern.platformdirs: 2.6.2\n",
      "platform: 1.0.8\n",
      "platformdirs: 3.10.0\n",
      "platformdirs.version: 3.10.0\n",
      "prompt_toolkit: 3.0.42\n",
      "psutil: 5.9.8\n",
      "pure_eval: 0.2.2\n",
      "pure_eval.version: 0.2.2\n",
      "pydevd: 2.9.5\n",
      "pygments: 2.18.0\n",
      "pyparsing: 3.1.2\n",
      "pytz: 2024.1\n",
      "re: 2.2.1\n",
      "requests: 2.31.0\n",
      "requests.__version__: 2.31.0\n",
      "requests.packages.idna: 3.4\n",
      "requests.packages.idna.idnadata: 15.0.0\n",
      "requests.packages.idna.package_data: 3.4\n",
      "requests.packages.urllib3: 2.1.0\n",
      "requests.packages.urllib3._version: 2.1.0\n",
      "requests.packages.urllib3.connection: 2.1.0\n",
      "requests.packages.urllib3.util.ssl_match_hostname: 3.5.0.1\n",
      "requests.utils: 2.31.0\n",
      "six: 1.16.0\n",
      "socketserver: 0.4\n",
      "socks: 1.7.1\n",
      "stack_data: 0.6.2\n",
      "stack_data.version: 0.6.2\n",
      "tqdm: 4.65.0\n",
      "tqdm._dist_ver: 4.65.0\n",
      "tqdm.cli: 4.65.0\n",
      "tqdm.version: 4.65.0\n",
      "traitlets: 5.14.3\n",
      "traitlets._version: 5.14.3\n",
      "urllib.request: 3.12\n",
      "urllib3: 2.1.0\n",
      "urllib3._version: 2.1.0\n",
      "urllib3.connection: 2.1.0\n",
      "urllib3.util.ssl_match_hostname: 3.5.0.1\n",
      "wcwidth: 0.2.13\n",
      "xmlrpc.client: 3.12\n",
      "zlib: 1.0\n",
      "zmq: 25.1.1\n",
      "zmq.sugar: 25.1.1\n",
      "zmq.sugar.version: 25.1.1\n",
      "zstandard: 0.19.0\n",
      "zstandard.backend_c: 0.19.0\n"
     ]
    }
   ],
   "source": [
    "# Display environment detail\n",
    "environment_info()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "97f97021-a9c2-4184-bd54-f186abc49305",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████| 1168/1168 [11:45<00:00,  1.66it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File saved successfully at ./data/TCGA_GeneExpression/uterus/clinical_gdc_manifest.txt\n",
      "Clinical manifest shape: (1168, 102)\n",
      "Cleaned clinical manifest shape: (1168, 44)\n",
      "No directories to clean. Cleaning done.\n",
      "Gene expression shape: (1168, 60660)\n",
      "Gene key shape: (60660, 2)\n",
      "{'percent_unique_filenames_in_manifest': 100.0, 'percent_overlap_with_directory': 100.0, 'percent_non_overlap_with_directory': 0.0}\n"
     ]
    }
   ],
   "source": [
    "# Specify the cancer type to analyze \n",
    "# Analyzed in manuscript (\"uterus\", \"thyroid\", \"lung\", \"kidney\", \"breast\", \"bone_marrow_and_blood\")\n",
    "cancer = \"uterus\"\n",
    "\n",
    "# --- Prerequisites ---\n",
    "# Before running this script, ensure the following:\n",
    "# 1. A directory structure exists: './data/TCGA_GeneExpression/{cancer}/'\n",
    "#    (e.g., './data/TCGA_GeneExpression/uterus/').\n",
    "# 2. Within the above directory, the following must be present:\n",
    "#    - A GDC manifest txt file with 'gdc_manifest' in its name. This file must be downloaded\n",
    "#      from the GDC Data Portal and placed in './data/TCGA_GeneExpression/{cancer}/'.\n",
    "#    - A subdirectory named 'gene_expression' containing all downloaded gene expression `.tsv` files\n",
    "#      (also from the GDC Data Portal) for the specified cancer type.  \n",
    "#      (select cancer type filter > transcriptome profiling > Gene Expression Quantification > RNA-Seq > '.tsv' files)\n",
    "\n",
    "\n",
    "# --- Step 1: Load the manifest file from GDC ---\n",
    "# This step loads the manifest file listing metadata about available files for the specified cancer type.\n",
    "gdc_manifest = load_manifest_file(f'./data/TCGA_GeneExpression/{cancer}/')\n",
    "\n",
    "# --- Step 2: Fetch clinical data using the GDC API ---\n",
    "# This step queries the GDC API to fetch clinical data such as diagnosis details and sample types.\n",
    "# Ensure an active internet connection to access the GDC API.\n",
    "clinical_manifest = fetch_clinical_data_for_manifest(gdc_manifest, n=None, t=0)\n",
    "\n",
    "# Save the clinical data for reuse. This file will be saved as 'clinical_gdc_manifest.txt'.\n",
    "save_clinical_manifest(clinical_manifest, f'./data/TCGA_GeneExpression/{cancer}/')\n",
    "\n",
    "# --- Step 3: Load and clean the clinical manifest ---\n",
    "# Load the saved clinical manifest file and clean it by removing irrelevant or empty columns.\n",
    "clinical_manifest = load_clinical_manifest(f'./data/TCGA_GeneExpression/{cancer}/')\n",
    "clinical_manifest = clean_dataframe(clinical_manifest)\n",
    "\n",
    "# --- Step 4: Organize and compile gene expression data ---\n",
    "# Ensure that the subdirectory './data/TCGA_GeneExpression/{cancer}/gene_expression/' exists\n",
    "# and contains the downloaded gene expression `.tsv` files from the GDC Data Portal.\n",
    "\n",
    "# Move all `.tsv` files from subdirectories into the 'gene_expression' directory.\n",
    "organize_tsv_files(f'./data/TCGA_GeneExpression/{cancer}/gene_expression')\n",
    "\n",
    "# Load the gene expression data from the `.tsv` files, extracting the specified feature ('fpkm_unstranded').\n",
    "gene_exps, gene_key = load_gene_expression(\n",
    "    clinical_manifest, f'./data/TCGA_GeneExpression/{cancer}/gene_expression', 'fpkm_unstranded'\n",
    ")\n",
    "\n",
    "# Save the processed gene expression and gene key data for later use.\n",
    "save_dataframes(gene_exps, gene_key, cancer)\n",
    "\n",
    "# --- Step 5: Analyze and verify filenames ---\n",
    "# Compare filenames in the clinical manifest with those in the 'gene_expression' directory\n",
    "# to identify any mismatches or missing files.\n",
    "print(analyze_filenames(clinical_manifest, f'./data/TCGA_GeneExpression/{cancer}/gene_expression'))\n",
    "\n",
    "# --- Step 6: Reload saved gene expression and gene key data ---\n",
    "# Reload the saved gene expression and gene key data from the CSV files.\n",
    "# This is useful for reanalysis or downstream processing without repeating earlier steps.\n",
    "gene_exps, gene_key = load_dataframes(\n",
    "    f'./data/TCGA_GeneExpression/{cancer}/gene_expression.csv', \n",
    "    f'./data/TCGA_GeneExpression/{cancer}/gene_key.csv'\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6c57748a-9b5b-4abe-a799-78141bdcba3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003.15</td>\n",
       "      <td>TSPAN6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005.6</td>\n",
       "      <td>TNMD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419.13</td>\n",
       "      <td>DPM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457.14</td>\n",
       "      <td>SCYL3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>C1orf112</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id gene_name\n",
       "0  ENSG00000000003.15    TSPAN6\n",
       "1   ENSG00000000005.6      TNMD\n",
       "2  ENSG00000000419.13      DPM1\n",
       "3  ENSG00000000457.14     SCYL3\n",
       "4  ENSG00000000460.17  C1orf112"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_key.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4725615b-ccea-4bcd-80b5-35085b3f24e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000000003.15</th>\n",
       "      <th>ENSG00000000005.6</th>\n",
       "      <th>ENSG00000000419.13</th>\n",
       "      <th>ENSG00000000457.14</th>\n",
       "      <th>ENSG00000000460.17</th>\n",
       "      <th>ENSG00000000938.13</th>\n",
       "      <th>ENSG00000000971.16</th>\n",
       "      <th>ENSG00000001036.14</th>\n",
       "      <th>ENSG00000001084.13</th>\n",
       "      <th>ENSG00000001167.14</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000288661.1</th>\n",
       "      <th>ENSG00000288662.1</th>\n",
       "      <th>ENSG00000288663.1</th>\n",
       "      <th>ENSG00000288665.1</th>\n",
       "      <th>ENSG00000288667.1</th>\n",
       "      <th>ENSG00000288669.1</th>\n",
       "      <th>ENSG00000288670.1</th>\n",
       "      <th>ENSG00000288671.1</th>\n",
       "      <th>ENSG00000288674.1</th>\n",
       "      <th>ENSG00000288675.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>filename</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <td>19.2975</td>\n",
       "      <td>1.4941</td>\n",
       "      <td>12.5086</td>\n",
       "      <td>6.7530</td>\n",
       "      <td>3.9497</td>\n",
       "      <td>1.4546</td>\n",
       "      <td>16.7088</td>\n",
       "      <td>12.2630</td>\n",
       "      <td>20.4519</td>\n",
       "      <td>16.5588</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36.7550</td>\n",
       "      <td>0.8335</td>\n",
       "      <td>0.0</td>\n",
       "      <td>44.4953</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.2613</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0176</td>\n",
       "      <td>1.0098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <td>10.6894</td>\n",
       "      <td>0.3790</td>\n",
       "      <td>35.1229</td>\n",
       "      <td>5.4748</td>\n",
       "      <td>2.5248</td>\n",
       "      <td>3.5777</td>\n",
       "      <td>3.1174</td>\n",
       "      <td>8.7109</td>\n",
       "      <td>4.1676</td>\n",
       "      <td>13.2483</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.3599</td>\n",
       "      <td>0.1395</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.5829</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.5882</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0673</td>\n",
       "      <td>0.0350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75073929-f7b1-4165-a786-60f733b962d9.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <td>17.7992</td>\n",
       "      <td>0.4563</td>\n",
       "      <td>26.4298</td>\n",
       "      <td>5.0877</td>\n",
       "      <td>2.7658</td>\n",
       "      <td>4.6093</td>\n",
       "      <td>29.4178</td>\n",
       "      <td>15.9268</td>\n",
       "      <td>8.3757</td>\n",
       "      <td>9.7255</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.6512</td>\n",
       "      <td>0.1094</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.9021</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.9262</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0389</td>\n",
       "      <td>0.3266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <td>10.4508</td>\n",
       "      <td>0.3030</td>\n",
       "      <td>18.1245</td>\n",
       "      <td>9.1775</td>\n",
       "      <td>5.8430</td>\n",
       "      <td>0.8595</td>\n",
       "      <td>11.2738</td>\n",
       "      <td>14.0382</td>\n",
       "      <td>4.9385</td>\n",
       "      <td>15.3237</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.9381</td>\n",
       "      <td>0.1635</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.6598</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.1973</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0387</td>\n",
       "      <td>0.2218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.augmented_star_gene_counts.tsv</th>\n",
       "      <td>27.1184</td>\n",
       "      <td>0.3620</td>\n",
       "      <td>26.7232</td>\n",
       "      <td>6.2870</td>\n",
       "      <td>3.4341</td>\n",
       "      <td>3.0606</td>\n",
       "      <td>55.9515</td>\n",
       "      <td>17.7718</td>\n",
       "      <td>20.5075</td>\n",
       "      <td>15.8846</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.4691</td>\n",
       "      <td>0.0762</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.8004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.2337</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0347</td>\n",
       "      <td>0.1391</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 60660 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    ENSG00000000003.15  \\\n",
       "filename                                                                 \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...             19.2975   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...             10.6894   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...             17.7992   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...             10.4508   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...             27.1184   \n",
       "\n",
       "                                                    ENSG00000000005.6  \\\n",
       "filename                                                                \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...             1.4941   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...             0.3790   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...             0.4563   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...             0.3030   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...             0.3620   \n",
       "\n",
       "                                                    ENSG00000000419.13  \\\n",
       "filename                                                                 \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...             12.5086   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...             35.1229   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...             26.4298   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...             18.1245   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...             26.7232   \n",
       "\n",
       "                                                    ENSG00000000457.14  \\\n",
       "filename                                                                 \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...              6.7530   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...              5.4748   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...              5.0877   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...              9.1775   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...              6.2870   \n",
       "\n",
       "                                                    ENSG00000000460.17  \\\n",
       "filename                                                                 \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...              3.9497   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...              2.5248   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...              2.7658   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...              5.8430   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...              3.4341   \n",
       "\n",
       "                                                    ENSG00000000938.13  \\\n",
       "filename                                                                 \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...              1.4546   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...              3.5777   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...              4.6093   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...              0.8595   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...              3.0606   \n",
       "\n",
       "                                                    ENSG00000000971.16  \\\n",
       "filename                                                                 \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...             16.7088   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...              3.1174   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...             29.4178   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...             11.2738   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...             55.9515   \n",
       "\n",
       "                                                    ENSG00000001036.14  \\\n",
       "filename                                                                 \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...             12.2630   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...              8.7109   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...             15.9268   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...             14.0382   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...             17.7718   \n",
       "\n",
       "                                                    ENSG00000001084.13  \\\n",
       "filename                                                                 \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...             20.4519   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...              4.1676   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...              8.3757   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...              4.9385   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...             20.5075   \n",
       "\n",
       "                                                    ENSG00000001167.14  ...  \\\n",
       "filename                                                                ...   \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...             16.5588  ...   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...             13.2483  ...   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...              9.7255  ...   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...             15.3237  ...   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...             15.8846  ...   \n",
       "\n",
       "                                                    ENSG00000288661.1  \\\n",
       "filename                                                                \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...                0.0   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...                0.0   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...                0.0   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...                0.0   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...                0.0   \n",
       "\n",
       "                                                    ENSG00000288662.1  \\\n",
       "filename                                                                \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...            36.7550   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...             7.3599   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...             6.6512   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...             9.9381   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...            14.4691   \n",
       "\n",
       "                                                    ENSG00000288663.1  \\\n",
       "filename                                                                \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...             0.8335   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...             0.1395   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...             0.1094   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...             0.1635   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...             0.0762   \n",
       "\n",
       "                                                    ENSG00000288665.1  \\\n",
       "filename                                                                \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...                0.0   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...                0.0   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...                0.0   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...                0.0   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...                0.0   \n",
       "\n",
       "                                                    ENSG00000288667.1  \\\n",
       "filename                                                                \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...            44.4953   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...            11.5829   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...            13.9021   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...            25.6598   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...            20.8004   \n",
       "\n",
       "                                                    ENSG00000288669.1  \\\n",
       "filename                                                                \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...                0.0   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...                0.0   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...                0.0   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...                0.0   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...                0.0   \n",
       "\n",
       "                                                    ENSG00000288670.1  \\\n",
       "filename                                                                \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...             8.2613   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...             5.5882   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...             2.9262   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...             5.1973   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...             3.2337   \n",
       "\n",
       "                                                    ENSG00000288671.1  \\\n",
       "filename                                                                \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...                0.0   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...                0.0   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...                0.0   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...                0.0   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...                0.0   \n",
       "\n",
       "                                                    ENSG00000288674.1  \\\n",
       "filename                                                                \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...             0.0176   \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...             0.0673   \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...             0.0389   \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...             0.0387   \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...             0.0347   \n",
       "\n",
       "                                                    ENSG00000288675.1  \n",
       "filename                                                               \n",
       "6a6ccdf5-9405-409a-a161-72e875400889.rna_seq.au...             1.0098  \n",
       "e5325dd8-867a-4cc2-8b93-9b0b58fc32f3.rna_seq.au...             0.0350  \n",
       "75073929-f7b1-4165-a786-60f733b962d9.rna_seq.au...             0.3266  \n",
       "119ac668-eab5-42e5-a7a5-f28720173235.rna_seq.au...             0.2218  \n",
       "09c2ef25-0496-4574-995b-d493f9ddf724.rna_seq.au...             0.1391  \n",
       "\n",
       "[5 rows x 60660 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_exps.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54857fca-ae13-4bc4-b483-6f16e7dbcc68",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
